src/query/graphite/common/context.go

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package common
+
+import (
+	ctx "context"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3/src/query/graphite/context"
+)
+
+// contextBase are the real content of a Context, minus the lock so that we
+// can safely copy a context without violating the rules of go vet.
+// nolint
+type contextBase struct {
+	// TimeRangeAdjusted is a boolean indicating whether the time range has an adjustment.
+	TimeRangeAdjusted bool
+
+	// The start time to query against.
+	StartTime time.Time
+
+	// The end time to query against.
+	EndTime time.Time
+
+	// TimeRangeAdjustment is the time range adjustment made to the query.
+	TimeRangeAdjustment TimeRangeAdjustment
+
+	// The underlying engine.
+	Engine QueryEngine
+
+	// Trace records traces.
+	Trace Tracer
+
+	// Timeout indicates whether to use a custom timeout when fetching data,
+	// specify zero to indicate default timeout or a positive value
+	Timeout time.Duration
+
+	parent          *Context
+	reqCtx          ctx.Context
+	storageContext  context.Context
+	state           map[string]interface{}
+	fetchBreakdowns []FetchBreakdown
+}
+
+// Context is the parameters to a query evaluation.
+type Context struct {
+	sync.RWMutex
+	contextBase
+}
+
+// ContextOptions provides the options to create the context with
+type ContextOptions struct {
+	Start     time.Time
+	End       time.Time
+	Engine    QueryEngine
+	LocalOnly bool
+	UseCache  bool
+	UseM3DB   bool
```

Comment:
Can remove LocalOnly, UseCache and UseM3DB here.

Reply:
Ah damn, removed them from regular context object but missed cleaning up contextBase